### PR TITLE
feat: Additional checks for metadata [DHIS2-14297]

### PR DESCRIPTION
**Pull request by jason-p-pickering**
- Unused category combinations
- Category combinations which have shared category options in their categories
- Category option combinations with the wrong cardinality
- Categories with the same category options
- Category options without categories
- Category option combinations with disjoint category options
- Category options part of two or more groups in a category option group set
- Potentially incomplete category option group sets
- Option sets with no options
- Unused option sets
- Option sets with incorrect sort order
- Unused visualizations
- Unused maps
- Unused dashboards
- Empty dashboards
- Aggregate data elements which do not belong to a data element group
- Abandoned aggregate data elements
- Aggregate data elements with inconsistent aggregation operators based on their type 
- Aggregate data elements with no analysis
- Duplicated indicator types (based on their factor)
- Indicators not present in any indicator groups
- Indicators not used by any analytical object
- Indicators which have exactly the same numerator and denominator
- Indicators which have the same terms in their numerators and denominators
- Periods more than three years in the future
- Periods more than 30 years in the past
- Distinct periods with the same start and end date
- Validation rules with an undefined missing value strategy
- Data element groups with less than two members
- Data elements which are part of two or more data element groups in a group set
-  Groups with less than two members 
- Program rule actions with no message template
- Program rules with no action
- Program rules with no priority

### CODE REVIEW

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityProgramRuleNoActionControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    private ProgramService programService;
+
+    @Autowired
+    private ProgramRuleService programRuleService;
+
+    @Autowired
+    private ProgramStageService programStageService;
+
+    @Autowired
+    ProgramRuleActionService programRuleActionService;
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityProgramRuleNoActionControllerTest.java:62]: private

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ * @author Jason P. Pickering
+ */
+class DataIntegrityVisualizationNotUsedOneYearControllerTest extends AbstractDataIntegrityIntegrationTest
+{
+
+    @Autowired
+    private DataStatisticsEventStore dataStatisticsEventStore;
+
+    @Autowired
+    VisualizationService visualizationService;
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java:58]: private

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    private static final String check = "visualizations_notviewed_1y";
+
+    private static final String viz_uid = "YngaQVeOC44";
+
+    @Test
+    void testUnusedVisualizationsExist()
+    {
+
+        SimpleDateFormat formatter = new SimpleDateFormat( "yyyy-MM-dd" );
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java:72]: avoid `SimpleDateFormat`, better use `java.time` - If I remember you just wanted a date in the past. You did use this in another test.

**Review by jbee — CHANGES_REQUESTED** [review]
Looks very nice overall. 
Great you extended and fixed the general issue assertion.

In the modelling of JSON objects there are some details to do differently (see line comments).

Tests look great. The REST tests should also allow testers to understand and execute the test scenario manually when wanted.

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/webapi/json/domain/JsonCategory.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        return getList( "categoryOptions", JsonCategoryOption.class );
+    }
+
+    /*
+     * default JsonList<JsonCategory> getCategoryCombos() { return getList(
+     * "categoryCombos", JsonCategoryCombo.class ); }
+     */
+
+    default String getDimensionType()
```
[the remark this answers is not in the document]
Comment (jason-p-pickering) [reply dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/webapi/json/domain/JsonCategory.java]: Fixed

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/webapi/json/domain/JsonCategoryCombination.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * @author Jason P. Pickering
+ */
+public interface JsonCategoryCombination extends JsonIdentifiableObject
+{
+    default JsonList<JsonCategory> getCategories()
+    {
+        return getList( "categories", JsonCategory.class );
+    }
+
+    default String getDimensionType()
```
[the remark this answers is not in the document]
Comment (jason-p-pickering) [reply dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/webapi/json/domain/JsonCategoryCombination.java]: Thanks. Fixed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityMapsNotUsedOneYearControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    private DataStatisticsEvent dse1;
+
+    private static final String check = "maps_notviewed_1y";
+
+    @Test
+    void testUnusedVisualizationsExist()
+    {
+
+        Date date = Date.from( ZonedDateTime.now().minusYears( 1 ).minusDays( 1 ).toInstant() );
```
[the remark this answers is not in the document]
Comment (jason-p-pickering) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityMapsNotUsedOneYearControllerTest.java]: Done

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityMultipleRootsControllerTest.java [hunk 7]
```diff
@@ -32,13 +32,17 @@
...
-import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+/**
+ * Test for multiple roots in the organisation unit hierarchy.
+ * {@see dhis-2/dhis-services/dhis-service-administration/src/main/resources/data-integrity-checks/orgunits/orgunits_multiple_roots.yaml}
+ *
+ * @author Jason P. Pickering
+ */
 class DataIntegrityMultipleRootsControllerTest extends AbstractDataIntegrityIntegrationTest
```
[the remark this answers is not in the document]
Comment (jason-p-pickering) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityMultipleRootsControllerTest.java]: Renamed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java (hunk 2 above)
Comment (jason-p-pickering) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java]: Danke. Fixed.

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java (hunk 3 above)
Comment (jason-p-pickering) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/dataintegrity/DataIntegrityVisualizationNotUsedOneYearControllerTest.java]: Re-implemented with java.time

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/webapi/json/domain/JsonOption.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    default Integer getSortOrder()
+    {
+        try
+        {
+            return Integer.parseInt( get( "sortOrder" ).toString() );
+        }
+        catch ( NumberFormatException e )
+        {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (jason-p-pickering) [reply dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/webapi/json/domain/JsonOption.java:52]: Thanks. FIxed.

**Review by jbee — APPROVED** [review]

**Review by olavpo — APPROVED** [review]

**Pull request merged**